src/lib/snarky_group_hash/snarky_unchecked_hash.ml

```diff
@@ -0,0 +1,342 @@
+    ; ((18, 3), "-63482146340076490752")
+    ; ((15, 5), "-101522561076541440")
+    ; ((12, 7), "69490543161600")
+    ; ((9, 9), "280657428480")
+    ; ((6, 11), "255430032")
+    ; ((3, 13), "100584")
+    ; ((0, 15), "15")
+    ; ((24, 0), "199571139166470771769344")
+    ; ((21, 2), "824674128787069304832")
+    ; ((18, 4), "-7951403445605351424")
+    ; ((15, 6), "-37420516674680832")
+    ; ((12, 8), "-66000709716480")
+    ; ((9, 10), "-61039617408")
+    ; ((6, 12), "-31603264")
+    ; ((3, 14), "-8712")
+    ; ((1, 16), "-1") ]
+
+  let d31_tuples =
+    [ ((0, 0), "-1")
+    ; ((0, 1), "5")
+    ; ((3, 0), "10536")
+    ; ((0, 2), "-10")
+    ; ((3, 1), "9480")
+    ; ((0, 3), "10")
+    ; ((6, 0), "4024944")
+    ; ((3, 2), "-4488")
+    ; ((0, 4), "-5")
+    ; ((6, 1), "2108304")
+    ; ((3, 3), "2904")
+    ; ((0, 5), "1") ]
+
+  let d32_tuples =
+    [ ((0, 0), "1")
+    ; ((0, 1), "-10")
+    ; ((3, 0), "12636")
+    ; ((0, 2), "45")
+    ; ((3, 1), "20256")
+    ; ((0, 3), "-120")
+    ; ((6, 0), "51578784")
+    ; ((3, 2), "-158448")
+    ; ((0, 4), "210")
+    ; ((6, 1), "426572352")
+    ; ((3, 3), "149472")
+    ; ((0, 5), "-252")
+    ; ((9, 0), "74892394368")
+    ; ((6, 2), "-178487712")
+    ; ((3, 4), "146472")
+    ; ((0, 6), "210")
+    ; ((9, 1), "42705805824")
+    ; ((6, 3), "-194173056")
+    ; ((3, 5), "-328224")
+    ; ((0, 7), "-120")
+    ; ((12, 0), "38682048607488")
+    ; ((9, 2), "217678171392")
+    ; ((6, 4), "339663456")
+    ; ((3, 6), "208656")
+    ; ((0, 8), "45")
+    ; ((12, 1), "-44449457564160")
+    ; ((9, 3), "-122450296320")
+    ; ((6, 5), "-126498240")
+    ; ((3, 7), "-58080")
+    ; ((0, 9), "10")
+    ; ((15, 0), "6454061238316032")
+    ; ((12, 2), "22224728782080")
+    ; ((9, 4), "30612574080")
+    ; ((6, 6), "21083040")
+    ; ((3, 8), "7260")
+    ; ((0, 10), "1") ]
+
+  let field_of_string s =
+    let s' =
+      if s.[0] = '-' then String.sub s ~pos:1 ~len:(String.length s - 1) else s
+    in
+    Sexp.of_string_conv_exn s' F.t_of_sexp
+
+  let field_map = List.map ~f:(fun (a, b) -> (a, field_of_string b))
+
+  let n1tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n1_tuples)
+
+  let d1tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d1_tuples)
+
+  let n2tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n2_tuples)
+
+  let d2tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d2_tuples)
+
+  let n3tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n3_tuples)
+
+  let d31tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d31_tuples)
+
+  let d32tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d32_tuples)
+
+  let sum (xs : F.t list) =
+    List.fold ~init:F.zero ~f:(fun x y -> F.Infix.(x + y)) xs
+
+  let product (xs : F.t list) =
+    List.fold ~init:F.one ~f:(fun x y -> F.Infix.(x * y)) xs
+
+  let powers (v : F.t) (k : int) : F.t list =
+    let rec make_powers acc prev x =
+      if x = k then acc
```

Comment:
Probably want to reverse it here. I would return an array though (to have O(1) accesses later) and use `Array.of_list_rev` to reverse it and convert it to an array in one go.